xwiki-pro-macros-api/src/main/java/com/xwiki/macros/script/TagListScriptingService.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.macros.script;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.script.service.ScriptService;
+import org.xwiki.stability.Unstable;
+
+/**
+ * Tools to manipulate the data needed for the tagList macro.
+ *
+ * @version $Id$
+ * @since 1.20.0
+ */
+@Component
+@Named("taglist")
+@Singleton
+@Unstable
+public class TagListScriptingService implements ScriptService
+{
+    private static final String DIGIT_BLOCK = "DIGITS";
+
+    private static final String SEPARATOR = ",";
+
+    private static final String QUOTES = "'";
+
+    /**
+     * Transforms the input of a user in the right format.
+     *
+     * @param userInput string with spaces
+     * @return returns the user input in the right format
+     */
+    public String parseSpaces(String userInput)
+    {
+        StringBuilder builder = new StringBuilder();
+        String[] spaces = userInput.split(SEPARATOR);
+        for (int i = 0; i < spaces.length; i++) {
+            String space = spaces[i];
+            builder.append(QUOTES);
+            builder.append(space.trim());
+            builder.append(QUOTES);
+            if (i < spaces.length - 1) {
+                builder.append(SEPARATOR);
+            }
+        }
+        return builder.toString();
+    }
+
+    /**
+     * Handles the separation of tags into bins.
+     *
+     * @param tags list of tags
+     * @return the bins formed by the tags
+     */
+    public Map<String, List<String>> getBins(List<String> tags)
+    {
+
+        // The tags list will always be sorted because it is created from a TreeSet.
+        Map<String, List<String>> finalBins = new TreeMap<>();
+        StringBuilder key = new StringBuilder();
+        char binSeparator = '-';
+
+        // Limit the number of entries in a bin
+        int totalEntries = tags.size();
+        int binLimit = (int) (2 * Math.round(Math.pow(totalEntries, 1.0 / 3.0)));
```

Comment:
We want to split the tags into multiple bins and determine the right size for the bins dynamically without hardcoding the value. To do this, we need a formula, and I've chosen to use the Rice rule since it is the simplest. More info: https://medium.com/@maxmarkovvision/optimal-number-of-bins-for-histograms-3d7c48086fde